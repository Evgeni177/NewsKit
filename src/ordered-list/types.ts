import {MQ} from '../utils/style';

export interface OrderedListProps {
  overrides?: {
    spaceInline?: MQ<string>;
    content?: {
      stylePreset?: MQ<string>;
      typographyPreset?: MQ<string>;
    };
    counter?: {
      stylePreset?: MQ<string>;
      typographyPreset?: MQ<string>;
      minWidth?: MQ<string>;
    };
  };
}
