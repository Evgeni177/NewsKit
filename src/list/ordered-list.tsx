import React from 'react';
import {styled, getSizingFromTheme} from '../utils/style';
import {getBuiId} from '../utils/get-bui-id';
import {isValidNode} from '../utils/component';
import {StyledOl} from './styled';

const ListItem = styled.li`
  padding-left: ${getSizingFromTheme('sizing010')};
  margin-bottom: ${getSizingFromTheme('sizing040')};
  margin-left: ${getSizingFromTheme('sizing050')};
`;

const List = styled(StyledOl)`
  display: block;
`;

export const OrderedList: React.FC<{}> = ({children}) => (
  <List>
    {React.Children.map(children, node =>
      isValidNode(node) ? <ListItem key={getBuiId()}>{node}</ListItem> : null,
    )}
  </List>
);
