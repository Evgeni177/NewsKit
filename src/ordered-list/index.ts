export type {OrderedListProps} from './types';
export * from './ordered-list';
