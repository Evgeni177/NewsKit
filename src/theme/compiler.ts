/* eslint-disable no-use-before-define */
/* eslint-disable @typescript-eslint/no-explicit-any */
import {Theme, UncompiledTheme} from './types';
import {get} from '../utils/get';

const parseTokens = (str: unknown) => {
  if (typeof str !== 'string') {
    return [];
  }

  const result = [];
  let res: RegExpExecArray | null;

  const regex = /{{([a-z0-9.]+)}}/gi;
  // eslint-disable-next-line no-cond-assign
  while ((res = regex.exec(str))) {
    result.push(res[1]);
  }

  return result;
};

const parseAndGet = (
  theme: UncompiledTheme,
  value: unknown,
  stack: unknown[] = [],
): unknown => {
  if (stack.length > 4) {
    // eslint-disable-next-line no-console
    console.error(
      `Recursive loop detected, token stack: "${stack.join('", "')}"!`,
    );
    return '';
  }

  const tokens = parseTokens(value);
  if (tokens.length) {
    return tokens.reduce(
      (result: any, tokenPath) => {
        let tokenValue = get(theme, tokenPath);
        if (typeof tokenValue === 'undefined') {
          // eslint-disable-next-line no-console
          console.error(
            `Theme token "${tokenPath}" was not found when compiling theme!`,
          );
        } else {
          // We recurse down with that token value to support things like colors
          // e.g. (border color = semanticNegative010 = red010 = #ff0000).
          tokenValue = parseAndGet(theme, tokenValue, [tokenPath, ...stack]);
        }

        return tokens.length > 1
          ? result.replace(`{{${tokenPath}}}`, tokenValue as string)
          : recurseUnknown(theme, tokenValue);
      },
      value as string,
    );
  }

  // Not a token, return as is (e.g. hex code or px value).
  return value;
};

const recurseUnknown = (theme: UncompiledTheme, value: unknown): unknown => {
  if (typeof value === 'function') {
    // Evaluation function, e.g. for calculating line height, run function and pass it the theme.
    const fResult = value(theme);
    // Result could be an object (with tokens), so recursively compile.
    return recurseUnknown(theme, fResult);
  }
  if (typeof value === 'object' && !Array.isArray(value)) {
    // Nested object, recurse down
    return Object.entries(value as any).reduce(
      (acc, [k, v]: any) => {
        // Key could be a token, e.g. cropAdjustments fonts object
        const key = parseAndGet(theme, k) as string;
        acc[key] = recurseUnknown(theme, v);
        return acc;
      },
      {} as any,
    );
  }
  // Anything else, probably string, parse and get the token value
  return parseAndGet(theme, value);
};

export const compileTheme = (theme: UncompiledTheme | Theme): Theme => {
  if (!theme || theme.compiled === true) {
    return theme;
  }

  const {icons = {}, ...filteredTheme} = theme;
  return {
    ...(recurseUnknown(theme, filteredTheme) as any),
    icons,
    compiled: true,
  };
};
