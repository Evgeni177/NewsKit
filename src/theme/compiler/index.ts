/* eslint-disable no-use-before-define */
/* eslint-disable @typescript-eslint/no-explicit-any */
import {ThemeLoggerFunction, Theme, UncompiledTheme} from '../types';
import {get} from '../../utils/get';
import {ThemeCompilerOptions} from './types';

const parseTokens = (str: unknown) => {
  if (typeof str !== 'string') {
    return [];
  }

  const result = [];
  let res: RegExpExecArray | null;

  const regex = /{{([a-z0-9.]+)}}/gi;
  // eslint-disable-next-line no-cond-assign
  while ((res = regex.exec(str))) {
    result.push(res[1]);
  }

  return result;
};

const parseAndGet = (
  theme: UncompiledTheme,
  value: unknown,
  errorLogger: ThemeLoggerFunction,
  stack: unknown[] = [],
): unknown => {
  if (stack.length > 4) {
    errorLogger(
      `Recursive loop detected, token stack: "${stack.join('", "')}"!`,
    );
    return '';
  }

  const tokens = parseTokens(value);
  if (tokens.length) {
    return tokens.reduce(
      (result: any, tokenPath) => {
        let tokenValue = get(theme, tokenPath);
        if (typeof tokenValue === 'undefined') {
          errorLogger(
            `Theme token "${tokenPath}" was not found when compiling theme!`,
          );
        } else {
          // We recurse down with that token value to support things like colors
          // e.g. (border color = interactiveNegative010 = red010 = #ff0000).
          tokenValue = parseAndGet(theme, tokenValue, errorLogger, [
            tokenPath,
            ...stack,
          ]);
        }

        return tokens.length > 1
          ? result.replace(`{{${tokenPath}}}`, tokenValue as string)
          : recurseUnknown(theme, tokenValue, errorLogger);
      },
      value as string,
    );
  }

  // Not a token, return as is (e.g. hex code or px value).
  return value;
};

const recurseUnknown = (
  theme: UncompiledTheme,
  value: unknown,
  errorLogger: ThemeLoggerFunction,
): unknown => {
  if (typeof value === 'function') {
    // Evaluation function, e.g. for calculating line height, run function and pass it the theme.
    const fResult = value(theme);
    // Result could be an object (with tokens), so recursively compile.
    return recurseUnknown(theme, fResult, errorLogger);
  }
  if (typeof value === 'object' && !Array.isArray(value)) {
    // Nested object, recurse down
    return Object.entries(value as any).reduce(
      (acc, [k, v]: any) => {
        // Key could be a token, e.g. cropAdjustments fonts object
        const key = parseAndGet(theme, k, errorLogger) as string;
        acc[key] = recurseUnknown(theme, v, errorLogger);
        return acc;
      },
      {} as any,
    );
  }
  // Anything else, probably string, parse and get the token value
  return parseAndGet(theme, value, errorLogger);
};

export const compileTheme = (
  theme: UncompiledTheme | Theme,
  options: ThemeCompilerOptions = {},
): Theme => {
  if (!theme || theme.compiled === true) {
    return theme;
  }

  // eslint-disable-next-line no-console
  const errorLogger = options.errorLogger || console.error.bind(console);

  const {icons = {}, ...filteredTheme} = theme;
  return {
    ...(recurseUnknown(theme, filteredTheme, errorLogger) as any),
    icons,
    compiled: true,
  };
};
