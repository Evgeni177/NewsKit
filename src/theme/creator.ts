import {deepMerge} from '../utils/deep-merge';

import {ThemeOverrides, UncompiledTheme, ThemeBase} from './types';

import * as foundations from './foundations';
import * as presets from './presets';

import {componentDefaults} from './component-defaults';

interface CreateThemeArgs {
  name?: string;
  baseTheme?: UncompiledTheme;
  overrides?: ThemeOverrides;
}

export const createTheme = ({
  name = 'unnamed-newskit-theme',
  baseTheme,
  overrides = {},
}: CreateThemeArgs): UncompiledTheme => {
  if (baseTheme && baseTheme.compiled) {
    throw new Error(
      'createTheme received a compiled baseTheme. Base themes must be uncompiled.',
    );
  }
  return deepMerge(
    {
      themeVersion: 1,
      ...foundations,
      ...presets,
      componentDefaults,
      icons: {},
    } as ThemeBase,
    baseTheme,
    overrides,
    {name},
  );
};
