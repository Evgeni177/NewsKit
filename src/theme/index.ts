export * from './compiler';
export * from './creator';
export * from './emotion';
export * from './newskit-dark';
export * from './newskit-light';
export * from './types';
