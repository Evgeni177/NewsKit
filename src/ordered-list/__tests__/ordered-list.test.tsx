import {renderToFragmentWithTheme} from '../../test/test-utils';
import {createTheme} from '../../theme';
import {OrderedList} from '../ordered-list';

const myCustomTheme = createTheme({
  name: 'my-custom-ordered-list',
  overrides: {
    stylePresets: {
      customOrderedListItemCounter: {
        base: {
          color: '{{colors.blue060}}',
        },
      },
      customOrderedListItemContent: {
        base: {
          color: '{{colors.red060}}',
        },
      },
    },
  },
});

describe('OrderedList', () => {
  test('renders with defaults', () => {
    const props = {
      children: ['first node', 'second node', 'third node'],
    };
    const fragment = renderToFragmentWithTheme(OrderedList, props);
    expect(fragment).toMatchSnapshot();
  });

  test('renders with invalid nodes', () => {
    const props = {
      children: [null, 'second node', 'third node'],
    };
    const fragment = renderToFragmentWithTheme(OrderedList, props);
    expect(fragment).toMatchSnapshot();
  });

  test('renders with overrides', () => {
    const props = {
      children: ['first node', 'second node', 'third node'],
      overrides: {
        spaceInline: 'sizing010',
        content: {
          stylePreset: 'customOrderedListItemContent',
          typographyPreset: 'editorialParagraph030',
        },
        counter: {
          stylePreset: 'customOrderedListItemCounter',
          typographyPreset: 'editorialParagraph020',
          minWidth: 'sizing030',
        },
      },
    };
    const fragment = renderToFragmentWithTheme(
      OrderedList,
      props,
      myCustomTheme,
    );
    expect(fragment).toMatchSnapshot();
  });
});
