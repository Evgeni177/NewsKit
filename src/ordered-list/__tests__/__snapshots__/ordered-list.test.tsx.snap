// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderedList renders with defaults 1`] = `
<DocumentFragment>
  .emotion-0 {
  margin: 0;
  padding: 0;
  list-style: none;
  counter-reset: item-counter;
}

.emotion-1 {
  margin-bottom: 16px;
  counter-increment: item-counter;
  color: #2e2e2e;
  font-family: "Noto Sans",sans-serif;
  font-size: 14px;
  line-height: 1.4285714285714286;
  font-weight: 400;
  letter-spacing: 0;
}

.emotion-1 svg {
  fill: #2e2e2e;
}

.emotion-1::before {
  content: '';
  display: block;
  height: 0;
  width: 0;
  margin-bottom: -0.32678571428571435em;
}

.emotion-1::after {
  content: '';
  display: block;
  height: 0;
  width: 0;
  margin-top: -0.4267857142857143em;
}

.emotion-1::before {
  content: counter(item-counter) '. ';
  color: #2e2e2e;
  font-family: "Noto Sans",sans-serif;
  font-size: 14px;
  line-height: 1.4285714285714286;
  font-weight: 400;
  letter-spacing: 0;
  display: inline-block;
  min-width: 24px;
}

.emotion-1::before svg {
  fill: #2e2e2e;
}

<ol
    class="emotion-0"
    role="list"
  >
    <li
      class="emotion-1"
    >
      first node
    </li>
    <li
      class="emotion-1"
    >
      second node
    </li>
    <li
      class="emotion-1"
    >
      third node
    </li>
  </ol>
</DocumentFragment>
`;

exports[`OrderedList renders with invalid nodes 1`] = `
<DocumentFragment>
  .emotion-0 {
  margin: 0;
  padding: 0;
  list-style: none;
  counter-reset: item-counter;
}

.emotion-1 {
  margin-bottom: 16px;
  counter-increment: item-counter;
  color: #2e2e2e;
  font-family: "Noto Sans",sans-serif;
  font-size: 14px;
  line-height: 1.4285714285714286;
  font-weight: 400;
  letter-spacing: 0;
}

.emotion-1 svg {
  fill: #2e2e2e;
}

.emotion-1::before {
  content: '';
  display: block;
  height: 0;
  width: 0;
  margin-bottom: -0.32678571428571435em;
}

.emotion-1::after {
  content: '';
  display: block;
  height: 0;
  width: 0;
  margin-top: -0.4267857142857143em;
}

.emotion-1::before {
  content: counter(item-counter) '. ';
  color: #2e2e2e;
  font-family: "Noto Sans",sans-serif;
  font-size: 14px;
  line-height: 1.4285714285714286;
  font-weight: 400;
  letter-spacing: 0;
  display: inline-block;
  min-width: 24px;
}

.emotion-1::before svg {
  fill: #2e2e2e;
}

<ol
    class="emotion-0"
    role="list"
  >
    <li
      class="emotion-1"
    >
      second node
    </li>
    <li
      class="emotion-1"
    >
      third node
    </li>
  </ol>
</DocumentFragment>
`;

exports[`OrderedList renders with overrides 1`] = `
<DocumentFragment>
  .emotion-0 {
  margin: 0;
  padding: 0;
  list-style: none;
  counter-reset: item-counter;
}

.emotion-1 {
  margin-bottom: 4px;
  counter-increment: item-counter;
  color: #ef1703;
  font-family: "Noto Sans",sans-serif;
  font-size: 18px;
  line-height: 1.5555555555555556;
  font-weight: 400;
  letter-spacing: 0;
}

.emotion-1::before {
  content: '';
  display: block;
  height: 0;
  width: 0;
  margin-bottom: -0.39027777777777783em;
}

.emotion-1::after {
  content: '';
  display: block;
  height: 0;
  width: 0;
  margin-top: -0.4902777777777778em;
}

.emotion-1::before {
  content: counter(item-counter) '. ';
  color: #0a68c1;
  font-family: "Noto Sans",sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
  letter-spacing: 0;
  display: inline-block;
  min-width: 12px;
}

<ol
    class="emotion-0"
    role="list"
  >
    <li
      class="emotion-1"
    >
      first node
    </li>
    <li
      class="emotion-1"
    >
      second node
    </li>
    <li
      class="emotion-1"
    >
      third node
    </li>
  </ol>
</DocumentFragment>
`;
